### PR TITLE
generate type alias for the `block` type

### PR DESCRIPTION
**Pull request by flier**
Consider the c/c++ code with [the clang `Block` type](http://clang.llvm.org/docs/BlockLanguageSpec.html).

```c++
// bindgen-flags: --block-extern-crate -- -fblocks

typedef unsigned long long size_t;

void atexit_b(void (^)(void));

typedef void *dispatch_data_t;
typedef bool (^dispatch_data_applier_t)(dispatch_data_t region,
                                        size_t offset,
                                        const void *buffer,
                                        size_t size);
bool dispatch_data_apply(dispatch_data_t data,
                         dispatch_data_applier_t applier); 
```
The generated code will use the [block](https://crates.io/crates/block) crate to wrap function/closure as a `block` pointer, instead of a `*const c_void` pointer without arguments/return type.
```rust
/* automatically generated by rust-bindgen */

#![allow(dead_code, non_snake_case, non_camel_case_types, non_upper_case_globals)]

#[macro_use]
extern crate block;
extern "C" {
    #[link_name = "\u{1}_Z8atexit_bU13block_pointerFvvE"]
    pub fn atexit_b(arg1: _bindgen_ty_id_19);
}
pub type dispatch_data_t = *mut ::std::os::raw::c_void;
pub type dispatch_data_applier_t = _bindgen_ty_id_26;
extern "C" {
    #[link_name = "\u{1}_Z19dispatch_data_applyPvU13block_pointerFbS_yPKvyE"]
    pub fn dispatch_data_apply(data: dispatch_data_t, applier: dispatch_data_applier_t) -> bool;
}
pub type _bindgen_ty_id_19 = *const ::block::Block<(), ()>;
pub type _bindgen_ty_id_26 =
    *const ::block::Block<(dispatch_data_t, usize, *const ::std::os::raw::c_void, usize), bool>;
```
Then, we can use [a simple wrapper](https://github.com/SSheldon/rust-dispatch/blob/e979ba2e7b8d52f2487516931e73f5d9736f16c4/src/data.rs#L144) to call those functions.

```rust
pub fn block<A, R, F>(closure: F) -> *const Block<A, R>
where
    A: BlockArguments,
    F: 'static + IntoConcreteBlock<A, Ret = R>,
{
    let block = ConcreteBlock::new(closure);
    let block = block.copy();
    let ptr = (&*block) as *const _;
    let ptr = unsafe { _Block_copy(ptr as *const c_void) };

    ptr as *const Block<A, R>
}

#[derive(Debug)]
pub struct Data {
    ptr: dispatch_data_t,
}

impl Data
    /// Traverses the memory of a dispatch data object and executes custom code on each region.
    pub fn apply<F>(&self, applier: F) -> bool
    where
        F: 'static + Fn(&Self, usize, &[u8]) -> bool,
    {
        let data = self.clone();
        let applier = block(
            move |ptr: dispatch_data_t, offset: usize, buffer: *const c_void, size: usize| {
                assert_eq!(data.ptr, ptr);

                let buf = unsafe { slice::from_raw_parts(buffer as *const u8, size) };

                applier(&data, offset, buf)
            },
        );

        unsafe { dispatch_data_apply(self.ptr, applier) }
    }
}

    #[test]
    fn test_data() {
         //...
        assert!(data.apply(move |_data: &Data, offset: usize, buf: &[u8]| {
            assert_eq!(offset, 0);
            assert_eq!(buf, b"lowor");

            c.set(buf.len());

            true
        }));
    }
```

### CODE REVIEW

**Comment by flier** [comment]
I agree the `*mut c_void` have better compatibility and less dependency, but I believe the block pointer with `c_void` is not a good practice, it is same to a callback function pointer without signature. 

If you prefer to keep the default generated code simple, I suggest to add a new option like `--block-signature` instead of clang's `-fblocks`, without the option will generate `*mut c_void` for the block pointer. But the behavior will be different to the `objc` options (`--objc-extern-crate -- -x objective-c`)

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -3560,6 +3598,49 @@ mod utils {
...
+            quote! {
+                #![allow(improper_ctypes)]
+
+                use block;
+            }
+        };
+
+        let void = helpers::ast_ty::raw_type(ctx, "c_void");
+        let def_block = quote! {
+            pub fn block<A, R, F>(closure: F) -> *const ::block::Block<A, R>
```
Comment (emilio) [inline src/codegen/mod.rs]: I think this should not really belong in bindgen, fwiw. This ties bindgen to whatever `block`s API is. And if something in that crate changes, then bindgen needs a new release.

I think this should live in the `block` crate, if anything.

File: src/codegen/mod.rs [hunk 2]
```diff
@@ -3871,4 +3952,48 @@ mod utils {
...
+
+            match *arg_ty.canonical_type(ctx).kind() {
+                TypeKind::Array(t, _) => {
+                    t.to_rust_ty_or_opaque(ctx, &())
+                        .to_ptr(ctx.resolve_type(t).is_const())
+                },
+                TypeKind::Pointer(inner) => {
+                    let inner = ctx.resolve_item(inner);
+                    let inner_ty = inner.expect_type();
+                    if let TypeKind::ObjCInterface(_) = *inner_ty.canonical_type(ctx).kind() {
```
Comment (emilio) [inline src/codegen/mod.rs]: Should this be in Type::to_rust_ty instead, on a pointer branch? Looks like it should to me, then this code would become much simpler. I don't think we really want to restrict this to blocks, right?

File: src/codegen/mod.rs [hunk 3]
```diff
@@ -636,9 +678,10 @@ impl CodeGenerator for Type {
 
                 // If this is a known named type, disallow generating anything
                 // for it too.
-                let spelling = self.name().expect("Unnamed alias?");
-                if utils::type_from_named(ctx, spelling).is_some() {
-                    return;
+                if let Some(spelling) = self.name() {
```
Comment (emilio) [inline src/codegen/mod.rs]: Hmm, we're not changing which `Alias` we generate, why is this change needed? It doesn't really make sense to have an unnamed alias.

File: src/codegen/mod.rs [hunk 4]
```diff
@@ -3071,23 +3114,25 @@ impl TryToRustTy for Type {
...
-                } else if let Some(ty) = utils::type_from_named(
-                    ctx,
-                    spelling,
-                )
-                {
-                    Ok(ty)
                 } else {
-                    utils::build_path(item, ctx)
+                    if let Some(spelling) = self.name() {
+                        if let Some(ty) = utils::type_from_named(ctx, spelling) {
```
Comment (emilio) [inline src/codegen/mod.rs]: I understand this can be needed, if we share this code between aliases and blocks. Do we have tests for anonymous blocks?

Also, this could use `and_then` to avoid duplicating the `else` branches. Something like:

```rust
} else if let Some(ty) = self.name().and_then(|name| utils::type_from_named(ctx, name)) {
    Ok(ty)
} else {
    utils::build_path(..)
}
```

Though I don't feel terribly strongly about it.

File: src/codegen/mod.rs (hunk 1 above)
Comment (flier) [reply src/codegen/mod.rs]: agree, its not necessary, I will remove it

File: src/codegen/mod.rs (hunk 2 above)
Comment (flier) [reply src/codegen/mod.rs]: agree, change to use 
```rust
arg_ty.try_to_rust_ty(ctx, &arg_item)
      .unwrap_or_else(|_| arg_item.to_rust_ty_or_opaque(ctx, &()))
```

File: src/codegen/mod.rs (hunk 3 above)
Comment (flier) [reply src/codegen/mod.rs]: sorry, forget to restore it before commit, my fault.

File: src/codegen/mod.rs (hunk 4 above)
Comment (flier) [reply src/codegen/mod.rs]: I'm not sure what's your meanings about anonymous blocks, but I test two cases, I guess most of function signature will use those.

```objc
void atexit_b(void (^)(void));

typedef void *dispatch_data_t;

typedef bool (^dispatch_data_applier_t)(dispatch_data_t region,
                                        size_t offset,
                                        const void *buffer,
                                        size_t size);

bool dispatch_data_apply(dispatch_data_t data,
                         dispatch_data_applier_t applier);
```

File: src/codegen/mod.rs [hunk 5]
```diff
@@ -610,6 +620,38 @@ impl CodeGenerator for Type {
...
+            TypeKind::BlockPointer(inner) => {
+                let inner_item = inner.into_resolver()
+                    .through_type_refs()
+                    .resolve(ctx);
+                let name = item.canonical_name(ctx);
+
+                let inner_rust_type = {
+                    // Its possible that we have better layout information than
+                    // the inner type does, so fall back to an opaque blob based
+                    // on our layout if converting the inner item fails.
```
Comment (emilio) [inline src/codegen/mod.rs]: This comment doesn't seem to make sense.

File: src/codegen/mod.rs [hunk 6]
```diff
@@ -3871,4 +3925,28 @@ mod utils {
 
         args
     }
+
+    pub fn fnsig_block(
+        ctx: &BindgenContext,
+        sig: &FunctionSig,
+    ) -> quote::Tokens {
+        let args = sig.argument_types().iter().map(|&(_, ty)| {
```
Comment (emilio) [inline src/codegen/mod.rs]: I think `ty.to_rust_ty_or_opaque(ctx, &())` would do just fine here, no need to `try_to_rust_ty`, unless you have a test-case in which it makes a difference.

File: src/codegen/mod.rs [hunk 7]
```diff
@@ -3560,6 +3591,29 @@ mod utils {
...
     }
 
+    pub fn prepend_block_header(
+        ctx: &BindgenContext,
+        result: &mut Vec<quote::Tokens>,
+    ) {
+        let use_block = if ctx.options().block_extern_crate {
+            quote! {
+                #![allow(improper_ctypes)]
+
```
Comment (emilio) [inline src/codegen/mod.rs]: nit: stray newline.

File: src/codegen/mod.rs [hunk 8]
```diff
@@ -3560,6 +3591,29 @@ mod utils {
...
+                extern crate block;
+            }
+        } else {
+            quote! {
+                #![allow(improper_ctypes)]
+
```
Comment (emilio) [inline src/codegen/mod.rs]: ditto regarding the two previous points.

File: src/codegen/mod.rs (hunk 5 above)
Comment (emilio) [reply src/codegen/mod.rs]: It makes me a bit uncomfortable to not have a switch to preserve the old behavior by default... But I won't block on that if you really think it's not worth it.

That being said, I think we should do it if it breaks any non-objc build :)

File: src/codegen/mod.rs [hunk 9]
```diff
@@ -610,6 +620,38 @@ impl CodeGenerator for Type {
...
+                let name = item.canonical_name(ctx);
+
+                let inner_rust_type = {
+                    // Its possible that we have better layout information than
+                    // the inner type does, so fall back to an opaque blob based
+                    // on our layout if converting the inner item fails.
+                    if let TypeKind::Function(fnsig) = inner_item.kind().expect_type().kind() {
+                        utils::fnsig_block(ctx, fnsig)
+                    } else {
+                        panic!("invalid block typedef")
```
Comment (emilio) [inline src/codegen/mod.rs]: It'd be nice to include in the message more diagnostics, like `inner_item` or something.

File: src/codegen/mod.rs [hunk 10]
```diff
@@ -3560,6 +3591,29 @@ mod utils {
         result.extend(old_items.into_iter());
     }
 
+    pub fn prepend_block_header(
+        ctx: &BindgenContext,
+        result: &mut Vec<quote::Tokens>,
+    ) {
+        let use_block = if ctx.options().block_extern_crate {
+            quote! {
+                #![allow(improper_ctypes)]
```
Comment (emilio) [inline src/codegen/mod.rs]: Please don't generate this attribute. There's no guarantee that bindings will be in a module on their own. We don't do it anywhere else.

File: src/codegen/mod.rs (hunk 5 above)
Comment (flier) [reply src/codegen/mod.rs]: I'm fine, we can add a new option like `--block-signature` instead of clang's `-fblocks` if you want, and use `*const c_void` without the option.

File: src/codegen/mod.rs (hunk 10 above)
Comment (emilio) [reply src/codegen/mod.rs]: Perhaps, but that won't get executed unless you enable C++ namespaces. I'd leave the responsibility of ignoring the warning to the bindgen consumer.

Comment (flier) [reply src/codegen/mod.rs]: ok

**Comment by flier** [comment]
@emilio do you mean write a new section like `Generating Bindings to Objective C`?

**Comment by flier** [comment]
Sure, I have submit a new issue #1386 to follow up it.

**Comment by emilio** [comment]
@bors-servo r+

Thanks!

**Comment by flier** [comment]
@emilio could you restart the test again?

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated

**Comment by emilio** [comment]
Yeah, looks like a fluke.

File: src/codegen/mod.rs [hunk 11]
```diff
@@ -3871,4 +3918,26 @@ mod utils {
...
+
+        let return_item = ctx.resolve_item(sig.return_type());
+        let ret_ty = if let TypeKind::Void = *return_item.kind().expect_type().kind() {
+            quote! { () }
+        } else {
+            return_item.to_rust_ty_or_opaque(ctx, &())
+        };
+
+        quote! {
+            *const ::block::Block<(#(#args),*), #ret_ty>
```
Comment (benw) [inline src/codegen/mod.rs]: Late to the party, but if there is a single arg then this expands to `Block<(arg), ret>` which parses the same as `Block<arg, ret>`. What's needed here is `Block<(arg,), ret>` so that args is a single-element tuple.

Comment (flier) [reply src/codegen/mod.rs]: agree, this should be a bug, I will submit a PR for this, thanks very much!
